tools/javadoc/javadoc.bzl

```diff
@@ -62,45 +61,76 @@ def _javadoc_library(ctx):
         #    to that directory, or a subdirectory, replacting dots with slashes.
-        javadoc_command += [
-            '-sourcepath $(find * -type d -name "*java" -print0 | tr "\\0" :)',
-            " ".join(ctx.attr.root_packages),
-            "-subpackages",
-            ":".join(ctx.attr.root_packages),
-        ]
+        jargs.add_all(ctx.attr.root_packages)
+        jargs.add_joined("-subpackages", ctx.attr.root_packages, join_with = ":")
     else:
         # Document exactly the code in the specified source files.
-        javadoc_command.append("$@")
-        args.add_all(ctx.files.srcs, map_each = _endswith_java)
+        jargs.add_all(ctx.files.srcs, map_each = _endswith_java)
 
     if ctx.attr.doctitle:
-        javadoc_command.append('-doctitle "%s"' % ctx.attr.doctitle)
+        jargs.add("-doctitle", ctx.attr.doctitle)
 
-    if ctx.attr.groups:
-        groups = []
-        for k, v in ctx.attr.groups.items():
-            groups.append("-group \"%s\" \"%s\"" % (k, ":".join(v)))
-        javadoc_command.append(" ".join(groups))
+    # Translate `groups` mapping to `-group k ":".join(v)`.
+    jargs.add_all(
+        ctx.attr.groups.items(),
+        map_each = _to_group,
+    )
 
-    if ctx.attr.exclude_packages:
-        javadoc_command.append("-exclude %s" % ":".join(ctx.attr.exclude_packages))
+    jargs.add_joined("-exclude", ctx.attr.exclude_packages, join_with = ":")
 
     for link in ctx.attr.external_javadoc_links:
-        javadoc_command.append("-linkoffline {0} {0}".format(link))
+        jargs.add("-linkoffline {0} {0}".format(link))
 
     if ctx.attr.bottom_text:
-        javadoc_command.append("-bottom '%s'" % ctx.attr.bottom_text)
-
-    # TODO(ronshapiro): Should we be using a different tool that doesn't include
-    # timestamp info?
-    jar_command = "%s/bin/jar cf %s -C %s ." % (java_home, ctx.outputs.jar.path, output_dir.path)
+        jargs.add("-bottom", ctx.attr.bottom_text)
 
     srcs = depset(transitive = [src.files for src in ctx.attr.srcs]).to_list()
+
+    # Invoke `javadoc` to generate Javadoc.
     ctx.actions.run_shell(
         inputs = srcs + classpath + ctx.files._jdk,
-        command = "%s && %s" % (" ".join(javadoc_command), jar_command),
-        outputs = [output_dir, ctx.outputs.jar],
-        arguments = [args],
+        outputs = [output_dir],
+        arguments = [
+            output_dir.path,
+            "true" if ctx.attr.root_packages else "false",
+            jargs,
+        ],
+        command = """
+outdir=$1; shift
+root_packages=$1; shift
+jargsfile=$1; shift
+
+# See also `TODO(b/167433657)` in javadoc.bzl.
+if $root_packages; then
+    sourcepath_args=(
+        -sourcepath
+        $(find * -type d -name '*.java' -print0 | tr '\\0' :)
+    )
+fi
+
+${JAVA_HOME}/bin/javadoc $jargsfile ${sourcepath_args[@]}
+""",
+        env = {
+            "JAVA_HOME": java_home,
+        },
+    )
+
+    # Zip the outputs using Bazel's deterministic `zipper` utility.
```

Comment:
Can we do this in a separate PR, just to keep this PR to a single purpose? With the old `jar` command, the shell script bit was much simpler because it supports `-C`, which is nice.

If a shell command gets to more than a couple of lines, I'd prefer to extract it out to a separate file. I think this may not be worth it, but we can talk about the value of using `zipper` separately.